File: crypto/src/main/java/com/blockstream/gdk/data/Address.kt
package com.blockstream.gdk.data

import android.os.Parcelable
import breez_sdk.LnInvoice
import breez_sdk.SwapInfo
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@Parcelize
data class Address constructor(
    @SerialName("address") val address: String,
    @SerialName("pointer") val pointer: Long = 0,
    @SerialName("address_type") val addressType: String? = null,
    @SerialName("branch") val branch: Long = 0,
    @SerialName("tx_count") val txCount: Long? = null,
    @SerialName("script") val script: String? = null,
    @SerialName("script_type") val scriptType: Int? = null,
    @SerialName("subaccount") val subaccount: Int? = null,
    @SerialName("subtype") val subType: Long? = null,
    @SerialName("user_path") val userPath: List<Long>? = null,
) : Parcelable {

    fun getUserPathAsInts(): List<Int>? {
        return userPath?.map { it.toInt() }
    }

    companion object{
        fun fromInvoice(invoice: LnInvoice): Address {
            return Address(address = invoice.bolt11)
        }

        fun fromSwapInfo(swapInfo: SwapInfo): Address {
            return Address(address = swapInfo.bitcoinAddress)
        }
    }
}

File: gdk/Models/Address.swift
import Foundation

import BreezSDK
import hw

public struct Address: Codable {

    enum CodingKeys: String, CodingKey {
        case address = "address"
        case pointer = "pointer"
        case branch = "branch"
        case userPath = "user_path"
        case subaccount = "subaccount"
        case scriptType = "script_type"
        case addressType = "address_type"
        case script = "script"
        case subtype = "subtype"
    }

    public let address: String?
    public let pointer: Int?
    public let branch: Int?
    public let subtype: UInt32?
    public let userPath: [UInt32]?
    public let subaccount: UInt32?
    public let scriptType: Int?
    public let addressType: String?
    public let script: String?

    public init(address: String? = nil, pointer: Int? = nil, branch: Int? = nil, subtype: UInt32? = nil, userPath: [UInt32]? = nil, subaccount: UInt32? = nil, scriptType: Int? = nil, addressType: String? = nil, script: String? = nil) {
        self.address = address
        self.pointer = pointer
        self.branch = branch
        self.subtype = subtype
        self.userPath = userPath
        self.subaccount = subaccount
        self.scriptType = scriptType
        self.addressType = addressType
        self.script = script
    }

    public static func from(invoice: LnInvoice) -> Address {
        return Address(address: invoice.bolt11)
    }

    public static func from(swapInfo: SwapInfo) -> Address {
        return Address(address: swapInfo.bitcoinAddress)
    }
}
